client/allocrunner/taskrunner/service_hook.go

```diff
@@ -219,7 +219,7 @@ func (h *serviceHook) Stop(ctx context.Context, req *interfaces.TaskStopRequest,
 
 func (h *serviceHook) getWorkloadServices() *serviceregistration.WorkloadServices {
 	// Interpolate with the task's environment
-	interpolatedServices := taskenv.InterpolateServices(h.taskEnv, h.services)
+	interpolatedServices := h.taskEnv.InterpolateServices(h.services)
```

Comment:
This shows a bit of the awkwardness of this approach -- unless the `TaskGroup` is in scope, we can't call `tg.InterpolatedServices`, so having the method indirection doesn't seem all that valuable.

Reply:
Wondering if it would make sense if we had something like

```
type Services []Service

(s Services) Interpolate(TaskEnv) Services
```

eliding the need for the `TaskGroup` reference? Seems like in practice you'd only get the Services from a tg anyway

Reply:
Ah nice, I think I will go with this approach. I was trying to have a `InterpoaltedService` type but it breaks quite a bit of things 😬

Reply:
It turns out that [`structs.Services` is already defined for something completely unrelated](https://github.com/hashicorp/nomad/blob/v1.5.0/nomad/structs/search.go#L24). I took this as a sign from the universe telling me to stop fiddling with this part of the code for now and just fix the bug 😅 

I'm going to close this one without merge and have updated https://github.com/hashicorp/nomad/pull/16402 with the right fix.

Thanks for all the help and apologies for the noise!

Reply:
heh FWIW all those constants should probably be named like `ServicesContext` instead of stealing the good names :slightly_smiling_face: